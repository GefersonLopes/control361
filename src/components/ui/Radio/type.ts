export type RadioProps = {
  id: string;
  name: string;
  value: string;
  label: string;
  checked?: boolean;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
};
