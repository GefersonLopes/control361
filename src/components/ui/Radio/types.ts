export interface RadioProps
  extends React.InputHTMLAttributes<HTMLInputElement> {
  id: string;
  name: string;
  value: string;
  label: React.ReactNode;
  checked?: boolean;
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  className?: string;
}
