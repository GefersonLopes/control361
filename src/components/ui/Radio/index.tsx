import clsx from "clsx";

import type { RadioProps } from "./type";

export default function Radio({
  id,
  name,
  value,
  label,
  checked,
  onChange,
  className,
}: RadioProps) {
  return (
    <label htmlFor={id} className="cursor-pointer flex items-center gap-4">
      <div className="grid place-items-center">
        <input
          type="radio"
          id={id}
          name={name}
          value={value}
          checked={checked}
          onChange={onChange}
          className={clsx(
            "peer col-start-1 row-start-1 appearance-none shrink-0 w-5 h-5 border-2",
            "border-[var(--color-primary)] rounded-full cursor-pointer",
            className,
          )}
        />

        <div
          className={clsx(
            "col-start-1 row-start-1 w-2.5 h-2.5 rounded-full",
            "bg-[var(--color-primary)] opacity-0 peer-checked:opacity-100 transition-opacity",
            className,
          )}
        />
      </div>

      <span className="text-[var(--color-light)] select-none font-inter text-medium text-xs">
        {label}
      </span>
    </label>
  );
}
