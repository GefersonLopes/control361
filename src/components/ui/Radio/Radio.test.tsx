import "@testing-library/jest-dom";

import { fireEvent, render, screen } from "@testing-library/react";

import Radio from "../Radio";

describe("Radio component", () => {
  const baseProps = {
    id: "radio-1",
    name: "group",
    value: "val1",
    label: "Opção 1",
    onChange: () => {},
  };

  it("renderiza o input radio com os atributos corretos e a classe sr-only", () => {
    render(<Radio {...baseProps} />);
    const input = screen.getByRole("radio", { name: /opção 1/i });
    expect(input).toBeInTheDocument();
    expect(input).toHaveAttribute("type", "radio");
    expect(input).toHaveAttribute("id", baseProps.id);
    expect(input).toHaveAttribute("name", baseProps.name);
    expect(input).toHaveAttribute("value", baseProps.value);
    expect(input).toHaveClass("sr-only");
  });

  it("renderiza o label apontando para o input e exibindo o texto", () => {
    render(<Radio {...baseProps} />);
    const label = screen.getByText(baseProps.label);
    expect(label).toBeInTheDocument();
    expect(label.closest("label")).toHaveAttribute("for", baseProps.id);
  });

  it("envolve tudo em um container com as classes flex e gap e aceita className extra", () => {
    render(<Radio {...baseProps} className="minha-classe" />);
    const input = screen.getByRole("radio", { name: /opção 1/i });
    const container = input.closest("div");
    expect(container).toHaveClass("flex", "items-center", "gap-4");
    expect(container).toHaveClass("minha-classe");
  });

  it("mostra o indicador interno com opacity-0 quando unchecked e opacity-100 quando checked", () => {
    const { rerender } = render(<Radio {...baseProps} checked={false} />);

    const input = screen.getByRole("radio", { name: /opção 1/i });
    const container = input.closest("div")!;
    const innerDot = container.querySelector("span > span")!;
    expect(innerDot).toHaveClass("opacity-0");

    rerender(<Radio {...baseProps} checked />);
    const innerDotChecked = screen
      .getByRole("radio", { name: /opção 1/i })
      .closest("div")!
      .querySelector("span > span")!;
    expect(innerDotChecked).toHaveClass("opacity-100");
  });

  it("propaga props arbitrários como data-* e aria-*", () => {
    render(
      <Radio
        {...baseProps}
        data-testid="radio-test"
        aria-label="Escolha uma opção"
      />,
    );
    const input = screen.getByTestId("radio-test");
    expect(input).toHaveAttribute("aria-label", "Escolha uma opção");
  });

  it("dispara onChange quando clicado", () => {
    const handleChange = jest.fn();
    render(<Radio {...baseProps} onChange={handleChange} />);
    const input = screen.getByRole("radio", { name: /opção 1/i });
    fireEvent.click(input);
    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  it("combina com o snapshot", () => {
    const { asFragment } = render(
      <Radio
        {...baseProps}
        checked
        className="snapshot-class"
        onChange={() => {}}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
