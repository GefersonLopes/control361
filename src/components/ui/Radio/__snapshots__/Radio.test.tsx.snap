// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio component combina com o snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex items-center gap-4 snapshot-class"
  >
    <input
      checked=""
      class="sr-only"
      id="radio-1"
      name="group"
      type="radio"
      value="val1"
    />
    <label
      class="flex items-center cursor-pointer select-none"
      for="radio-1"
    >
      <span
        aria-hidden="true"
        class="inline-flex items-center justify-center w-5 h-5 border-2 rounded-full border-[var(--color-primary)] transition-colors"
      >
        <span
          aria-hidden="true"
          class="w-2.5 h-2.5 rounded-full bg-[var(--color-primary)] opacity-100 transition-opacity"
        />
      </span>
      <span
        class="ml-2 text-xs font-inter text-[var(--color-light)]"
      >
        Opção 1
      </span>
    </label>
  </div>
</DocumentFragment>
`;
